### ISSUE TITLE
1.0.2 beta testing

### ISSUE THREAD

**Issue by aaron0browne** [chop-dbhi/data-models-validator#22]
Run on 21GB measurement file:

```
# time data-models-validator -model pedsnet -version 2.1.0 measurement.csv
Validating against model 'pedsnet/2.1.0'
* Evaluating 'measurement' table in 'measurement.csv'...
* Everything looks good!

real	32m15.840s
user	116m1.084s
sys	5m21.799s
```

Much better post-optimization!

**Comment by bruth** [comment]
That is great. Did you happen to observe the memory usage?

**Comment by aaron0browne** [comment]
No I didn't. I had top running but too many other things going on at the same time. Any good CLI for max memory usage like `time` that you know of?

**Comment by bruth** [comment]
I do not.

**Comment by aaron0browne** [comment]
The new line error truncation looks good too:

```
# time data-models-validator -model pedsnet -version 2.1.0 original_DRUG_EXPOSURE.csv:drug_exposure
Validating against model 'pedsnet/2.1.0'
* Evaluating 'drug_exposure' table in 'original_DRUG_EXPOSURE.csv'...
* Row-level issues were found.
+------+--------------------------------+-------------+--------------------------------+-----------------------------------------------------------------+
| CODE |             ERROR              | OCCURRENCES |             LINES              |                             EXAMPLE                             |
+------+--------------------------------+-------------+--------------------------------+-----------------------------------------------------------------+
|  203 | Value contains bare double     |        2012 | 4746325, 4746329, 4747955,     | line 4746325:                                                   |
|      | quotes (")                     |             | 4747957, 4747961, 4749461,     | `,"4122384","mg","735979","2009-06-04","2009-06-04              |
|      |                                |             | 4749477, 4750847, 4750861,     | 08:00:00","2212546","2009-06-01","2009-06-01                    |
|      |                                |             | 4750865 ... (2002 more)        | 16:33:23","2009-06-04","2009-06-04                              |
|      |                                |             |                                | 08:00:00","735979","risperidone 1 mg/tablet                     |
|      |                                |             |                                | UD "odt"","38000180","1.50000"," "," ","                        |
|      |                                |             |                                | ","100408","1800366","1.50000",,"4128794","PO","","","1847993"` |
|      |                                |             |                                | {column = 13}                                                   |
+------+--------------------------------+-------------+--------------------------------+-----------------------------------------------------------------+

real	14m40.670s
user	64m55.406s
sys	2m40.866s
```

**Comment by aaron0browne** [comment]
Are there any other changes I should test for 1.0.2?

**Comment by bruth** [comment]
I pushed a changed to not allow unquoted fields that are not empty (and [more tests](https://github.com/chop-dbhi/data-models-validator/blob/master/csv_test.go#L8-L41)). This now rejects the issue you mentioned yesterday about a trailing quote.

**Comment by aaron0browne** [comment]
With latest code:

```
# time data-models-validator -model pedsnet -version 2.1.0 measurement.csv
Validating against model 'pedsnet/2.1.0'
* Evaluating 'measurement' table in 'measurement.csv'...
* Problem reading CSV header: line 1: [code: 203] Value contains bare double quotes (")
{column = 1}

real	0m0.216s
user	0m0.022s
sys	0m0.012s
[root@reslnpedsndb02 2015-12-06T10:58]# head -n 1 measurement.csv
measurement_concept_id,measurement_date,measurement_id,measurement_order_date,measurement_order_time,measurement_result_date,measurement_result_time,measurement_source_concept_id,measurement_source_value,measurement_time,measurement_type_concept_id,operator_concept_id,person_id,provider_id,range_high,range_high_operator_concept_id,range_high_source_value,range_low,range_low_operator_concept_id,range_low_source_value,specimen_source_value,unit_concept_id,unit_source_value,value_as_concept_id,value_as_number,value_source_value,visit_occurrence_id
```

Obviously this header should be quoted, in order to meet the spec, but that error message doesn't seem right, either. I'll quote the header and re-run.

**Comment by aaron0browne** [comment]
The new version ran faster, somehow. Fantastic magical work!

```
# time data-models-validator -model pedsnet -version 2.1.0 new_measurement.csv:measurement
Validating against model 'pedsnet/2.1.0'
* Evaluating 'measurement' table in 'new_measurement.csv'...
* Everything looks good!

real	31m49.696s
user	112m29.831s
sys	5m9.267s
```

**Comment by aaron0browne** [comment]
Previous was slightly facetious...25s faster.

**Comment by aaron0browne** [comment]
There is something wrong here. This file passed the previous (optimization only) version, missing the quoting issue which is now being called out correctly, but there are many field-level issues that don't make sense...

```
time data-models-validator -model pedsnet -version 2.1.0 new_DRUG_EXPOSURE.CSV:drug_exposure
Validating against model 'pedsnet/2.1.0'
* Evaluating 'drug_exposure' table in 'new_DRUG_EXPOSURE.CSV'...
* Row-level issues were found.
+------+--------------------------------+-------------+--------------------------------+---------------------------------------------------------------------------------------------------+
| CODE |             ERROR              | OCCURRENCES |             LINES              |                                              EXAMPLE                                              |
+------+--------------------------------+-------------+--------------------------------+---------------------------------------------------------------------------------------------------+
|  203 | Value contains bare double     |    10530053 | 3, 5, 7, 9, 11, 13, 15, 17,    | line 3: `"92","3079155","40238246","2006-10-18 00:00:00","2007-01-05                              |
|      | quotes (")                     |             | 19, 21 ... (10530043 more)     | 00:00:00","2006-10-18 00:00:00","2007-01-05 00:00:00","2006-10-18 00:00:00",2006-10-18            |
|      |                                |             |                                | 00:00:00",38000177","Duplicate therapy",,"1",,"<XML><SIG>1 tab PO BID</SIG><QUANTITY>1 month      |
|      |                                |             |                                | supply</QUANTITY><REFILLS></REFILLS></XML>","4128794",,,"0",,"201530","100279286","72500010100610 |
|      |                                |             |                                | | DEPAKOTE 250 MG PO DR TABLET",,"Oral",,` {column = 9}                                           |
+------+--------------------------------+-------------+--------------------------------+---------------------------------------------------------------------------------------------------+
* Field-level issues were found.
+--------------------------+------+--------------------------------+-------------+--------------------------------+--------------------------------+
|          FIELD           | CODE |             ERROR              | OCCURRENCES |             LINES              |            SAMPLES             |
+--------------------------+------+--------------------------------+-------------+--------------------------------+--------------------------------+
| drug_exposure_order_time |  308 | Value is not a datetime        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 12634876:                 |
|                          |      | (YYYY-MM-DD HH:MM:SS)          |             | 18, 20 ... (10530044 more)     | `drug_exposure_order_time`     |
|                          |      |                                |             |                                | line 1467518:                  |
|                          |      |                                |             |                                | `drug_exposure_order_time`     |
|                          |      |                                |             |                                | line 8127448:                  |
|                          |      |                                |             |                                | `drug_exposure_order_time`     |
|                          |      |                                |             |                                | line 19381734:                 |
|                          |      |                                |             |                                | `drug_exposure_order_time`     |
|                          |      |                                |             |                                | line 16338862:                 |
|                          |      |                                |             |                                | `drug_exposure_order_time`     |
| drug_type_concept_id     |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 7740014:                  |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | `drug_type_concept_id`         |
|                          |      |                                |             |                                | line 1510050:                  |
|                          |      |                                |             |                                | `drug_type_concept_id`         |
|                          |      |                                |             |                                | line 19639204:                 |
|                          |      |                                |             |                                | `drug_type_concept_id`         |
|                          |      |                                |             |                                | line 14671682:                 |
|                          |      |                                |             |                                | `drug_type_concept_id`         |
|                          |      |                                |             |                                | line 13281864:                 |
|                          |      |                                |             |                                | `drug_type_concept_id`         |
| refills                  |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 18202546: `refills`       |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | line 13156290: `refills`       |
|                          |      |                                |             |                                | line 15196268: `refills`       |
|                          |      |                                |             |                                | line 16316520: `refills` line  |
|                          |      |                                |             |                                | 10351572: `refills`            |
| quantity                 |  306 | Value is not a number          |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 12033462: `quantity`      |
|                          |      | (float32)                      |             | 18, 20 ... (10530044 more)     | line 9216564: `quantity`       |
|                          |      |                                |             |                                | line 12543050: `quantity`      |
|                          |      |                                |             |                                | line 8050530: `quantity` line  |
|                          |      |                                |             |                                | 18853618: `quantity`           |
| days_supply              |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 16369724: `days_supply`   |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | line 12353416: `days_supply`   |
|                          |      |                                |             |                                | line 10080052: `days_supply`   |
|                          |      |                                |             |                                | line 2741242: `days_supply`    |
|                          |      |                                |             |                                | line 10006574: `days_supply`   |
| route_concept_id         |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 1850696:                  |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | `route_concept_id` line        |
|                          |      |                                |             |                                | 17419636: `route_concept_id`   |
|                          |      |                                |             |                                | line 16260278:                 |
|                          |      |                                |             |                                | `route_concept_id` line        |
|                          |      |                                |             |                                | 6091320: `route_concept_id`    |
|                          |      |                                |             |                                | line 17740380:                 |
|                          |      |                                |             |                                | `route_concept_id`             |
| effective_drug_dose      |  306 | Value is not a number          |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 18729476:                 |
|                          |      | (float32)                      |             | 18, 20 ... (10530044 more)     | `effective_drug_dose` line     |
|                          |      |                                |             |                                | 4288986: `effective_drug_dose` |
|                          |      |                                |             |                                | line 6410408:                  |
|                          |      |                                |             |                                | `effective_drug_dose` line     |
|                          |      |                                |             |                                | 1199968: `effective_drug_dose` |
|                          |      |                                |             |                                | line 2747744:                  |
|                          |      |                                |             |                                | `effective_drug_dose`          |
| dose_unit_concept_id     |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 10363466:                 |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | `dose_unit_concept_id`         |
|                          |      |                                |             |                                | line 7069048:                  |
|                          |      |                                |             |                                | `dose_unit_concept_id`         |
|                          |      |                                |             |                                | line 7394302:                  |
|                          |      |                                |             |                                | `dose_unit_concept_id`         |
|                          |      |                                |             |                                | line 20518510:                 |
|                          |      |                                |             |                                | `dose_unit_concept_id`         |
|                          |      |                                |             |                                | line 16815080:                 |
|                          |      |                                |             |                                | `dose_unit_concept_id`         |
| provider_id              |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 6255580: `provider_id`    |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | line 4356632: `provider_id`    |
|                          |      |                                |             |                                | line 14882650: `provider_id`   |
|                          |      |                                |             |                                | line 11102494: `provider_id`   |
|                          |      |                                |             |                                | line 17415040: `provider_id`   |
| visit_occurrence_id      |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 16554824:                 |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | `visit_occurrence_id`          |
|                          |      |                                |             |                                | line 12144584:                 |
|                          |      |                                |             |                                | `visit_occurrence_id`          |
|                          |      |                                |             |                                | line 12480964:                 |
|                          |      |                                |             |                                | `visit_occurrence_id` line     |
|                          |      |                                |             |                                | 8314724: `visit_occurrence_id` |
|                          |      |                                |             |                                | line 18422348:                 |
|                          |      |                                |             |                                | `visit_occurrence_id`          |
| drug_source_concept_id   |  305 | Value is not an integer        |    10530054 | 2, 4, 6, 8, 10, 12, 14, 16,    | line 4864510:                  |
|                          |      | (int32)                        |             | 18, 20 ... (10530044 more)     | `drug_source_concept_id`       |
|                          |      |                                |             |                                | line 3175792:                  |
|                          |      |                                |             |                                | `drug_source_concept_id`       |
|                          |      |                                |             |                                | line 7020918:                  |
|                          |      |                                |             |                                | `drug_source_concept_id`       |
|                          |      |                                |             |                                | line 18914174:                 |
|                          |      |                                |             |                                | `drug_source_concept_id`       |
|                          |      |                                |             |                                | line 5354116:                  |
|                          |      |                                |             |                                | `drug_source_concept_id`       |
+--------------------------+------+--------------------------------+-------------+--------------------------------+--------------------------------+

real	8m53.699s
user	45m31.377s
sys	3m27.431s
```

**Comment by bruth** [comment]
Indeed there is.

**Comment by bruth** [comment]
Probably a standard *off by one* error..

**Comment by bruth** [comment]
@aaron0browne This is a work in progress, but see #23 for a build that cleans a few bits up and includes tests to ensure the reader progresses.

**Comment by bruth** [comment]
Fixed in #23
